[SMTP] 이메일 보내기, 이메일 인증 확인하기 기능 구현

coolOlive opened the pull request:
## 🔥 Related Issues
resolved #28 

## 💜 작업 내용
- [x] 이메일 보내기
- [x] 이메일 인증번호 확인하기
 

## 👀 스크린샷 / GIF / 링크
![image](https://github.com/user-attachments/assets/42f92748-ce7f-45c5-9b9c-a9fd89829d84)

coolOlive committed:
feat: smtp 기능 구현

coolOlive committed:
refactor: 필요없는 파일 삭제 및 코드 수정

coolOlive committed:
refactor: smtp의 repository 구현부 수정 (JpaRepository를 jpql로 바꿈)

coolOlive committed:
refactor: 필요없는 파일 삭제 및 smtp의 exceptionHandler 추가

nangdev commented on src/main/java/com/sandbox/domain/smtp/entity/EmailVerification.java:
각 컬럼에 애노테이션 붙인거 섬세하시네요 근데 verified 필드는 뭐죠?

nangdev commented on src/main/java/com/sandbox/domain/smtp/repository/EmailVerificationRepository.java:
왜 리포지토리에 트랜잭셔너를 달아 놨을까 ㅎㅎㅎㅎㅎ
findByEmail을 왜 리스트로 받는거죠?
밑에도
save 수정도 가능할듯 엔티티에 생성할때 id필드 안넣어주면

coolOlive replied:
서비스로 이동시키겠습니다!

coolOlive replied:
-> 생성자 주입식으로 email, code입력받는거 하나 만들고 엔티티에서 세터 지우기 !

nangdev reviewed (approved):
# well done bro!

coolOlive commented on src/main/java/com/sandbox/domain/smtp/service/SMTPServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+        if (verification.isPresent()) {
+            EmailVerification emailVerification = verification.get();
+            emailVerification.setVerified(true);
+            emailVerificationRepository.save(emailVerification);
```
save하지 말고 delete해보자

coolOlive reviewed (commented):
리뷰 답변 완료!

coolOlive committed:
error: 리펙토링 중인데 repository에서 save안하려하니까 에러남

coolOlive committed:
refactor: service의 sendSecretNumber, authenticate 에서 repository 사용시 save, delete 활용 리팩토링

- sendSecretNumber: 원래 엔티티에서 setter를 사용해서 모든 값 변경 후 저장했는데, email이랑 code를 받는 생성자 만들고 이를 통해 객체를 save하는 형식으로 수정
> 기존 방식 : 중복검사 후 중복 있으면 set하고, 없으면 기본생성자로 된 객체 넘김
> 수정한 방식 : entity의 생성자 추가, 중복검사 후 중복 있으면 해당 객체 삭제 후 추가, 없으면 그냥 추가

- authenticate에서 save말고 delete를 호출하여 데이터베이스 사용을 효율적으로 하도록 함

coolOlive committed:
refactor: repository에서 사용하던 transactional를 service로 이동시킴

- sendSecretNumber는 @Transactional
- authenticate는 @Transactional(readOnly = true)

coolOlive committed:
refactoring: 변수, 함수명 정리 및 불필요한 import 제거

coolOlive committed:
refactoring: ErrorHandler, ErrorResp 활용하도록 수정

coolOlive committed:
refactoring: service의 sendSecretNumber에서 Optional에서 isPresent 처리방식 수정

- if (entity.isPresent())으로 검사한 구문을
entity.ifPresent(repo::delete); 로 수정
- 커밋할 때 노란 경고 뜨는거 해결한 것

nangdev commented on src/main/java/com/sandbox/domain/smtp/service/SMTPServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+        MimeMessage message = mailSender.createMimeMessage();
+        MimeMessageHelper helper = new MimeMessageHelper(message, true);
+
+        helper.setTo(to);
+        helper.setSubject("SANDBOX _ SMTP 인증 메일");
+        helper.setText("인증코드 : " + code);
```
빌드 메서드랑 샌드 메서드로 나누면 유지보수하기 용이할듯
